File: backend/utils/alphaVantageService.js
// Alpha Vantage API service
const axios = require('axios');
const logger = require('./logger');

const ALPHA_VANTAGE_BASE_URL = 'https://www.alphavantage.co/query';

class AlphaVantageService {
  constructor() {
    this.apiKey = process.env.ALPHA_VANTAGE_API_KEY;
    if (!this.apiKey) {
      logger.warn('ALPHA_VANTAGE_API_KEY not set in environment variables');
    }
  }

  async quote(symbol) {
    try {
      // Remove .NS or .BSE suffix for Alpha Vantage
      const cleanSymbol = symbol.replace(/\.(NS|BSE)$/i, '');
      
      const response = await axios.get(ALPHA_VANTAGE_BASE_URL, {
        params: {
          function: 'GLOBAL_QUOTE',
          symbol: cleanSymbol,
          apikey: this.apiKey
        }
      });

      const quote = response.data['Global Quote'];
      
      if (!quote || Object.keys(quote).length === 0) {
        throw new Error(`No data found for symbol: ${cleanSymbol}`);
      }

      // Map Alpha Vantage response to our format
      return {
        symbol: cleanSymbol,
        price: {
          symbol: cleanSymbol,
          regularMarketPrice: parseFloat(quote['05. price']) || 0,
          regularMarketPreviousClose: parseFloat(quote['08. previous close']) || 0,
          regularMarketVolume: parseInt(quote['06. volume']) || 0,
          regularMarketChange: parseFloat(quote['09. change']) || 0,
          regularMarketChangePercent: parseFloat((quote['10. change percent'] || '0').replace('%', '')) || 0,
          regularMarketOpen: parseFloat(quote['02. open']) || 0,
          regularMarketDayHigh: parseFloat(quote['03. high']) || 0,
          regularMarketDayLow: parseFloat(quote['04. low']) || 0,
          longName: cleanSymbol,
          shortName: cleanSymbol
        },
        regularMarketPrice: parseFloat(quote['05. price']) || 0,
        regularMarketPreviousClose: parseFloat(quote['08. previous close']) || 0,
        regularMarketVolume: parseInt(quote['06. volume']) || 0,
        longName: cleanSymbol,
        shortName: cleanSymbol
      };
    } catch (error) {
      logger.error(`Error fetching quote for ${symbol}:`, error.message);
      throw error;
    }
  }

  async getCompanyOverview(symbol) {
    try {
      const cleanSymbol = symbol.replace(/\.(NS|BSE)$/i, '');
      
      const response = await axios.get(ALPHA_VANTAGE_BASE_URL, {
        params: {
          function: 'OVERVIEW',
          symbol: cleanSymbol,
          apikey: this.apiKey
        }
      });

      const data = response.data;
      
      if (!data || Object.keys(data).length === 0 || data.Note) {
        // If rate limited or no data, return minimal info
        return {
          Symbol: cleanSymbol,
          Name: cleanSymbol,
          MarketCapitalization: 0,
          PERatio: null,
          DividendYield: null,
          '52WeekHigh': null,
          '52WeekLow': null,
          Beta: null,
          Sector: 'Unknown',
          Industry: 'Unknown'
        };
      }

      return data;
    } catch (error) {
      logger.error(`Error fetching company overview for ${symbol}:`, error.message);
      return {
        Symbol: symbol.replace(/\.(NS|BSE)$/i, ''),
        Name: symbol.replace(/\.(NS|BSE)$/i, ''),
        MarketCapitalization: 0,
        PERatio: null,
        DividendYield: null,
        '52WeekHigh': null,
        '52WeekLow': null,
        Beta: null,
        Sector: 'Unknown',
        Industry: 'Unknown'
      };
    }
  }

  async getDetailedQuote(symbol) {
    try {
      // Get both quote and overview for comprehensive data
      const [quote, overview] = await Promise.all([
        this.quote(symbol),
        this.getCompanyOverview(symbol)
      ]);

      // Merge the data
      return {
        symbol: quote.symbol,
        price: {
          ...quote.price,
          longName: overview.Name || quote.symbol,
          shortName: overview.Name || quote.symbol,
          marketCap: parseFloat(overview.MarketCapitalization) || 0,
          trailingPE: parseFloat(overview.PERatio) || null,
          beta: parseFloat(overview.Beta) || null
        },
        regularMarketPrice: quote.regularMarketPrice,
        regularMarketPreviousClose: quote.regularMarketPreviousClose,
        regularMarketVolume: quote.regularMarketVolume,
        longName: overview.Name || quote.symbol,
        shortName: overview.Name || quote.symbol,
        marketCap: parseFloat(overview.MarketCapitalization) || 0,
        trailingPE: parseFloat(overview.PERatio) || null,
        dividendYield: parseFloat(overview.DividendYield) || null,
        fiftyTwoWeekHigh: parseFloat(overview['52WeekHigh']) || null,
        fiftyTwoWeekLow: parseFloat(overview['52WeekLow']) || null,
        beta: parseFloat(overview.Beta) || null,
        sector: overview.Sector || 'Unknown',
        industry: overview.Industry || 'Unknown',
        summaryDetail: {
          trailingPE: parseFloat(overview.PERatio) || null,
          dividendYield: parseFloat(overview.DividendYield) || null,
          fiftyTwoWeekHigh: parseFloat(overview['52WeekHigh']) || null,
          fiftyTwoWeekLow: parseFloat(overview['52WeekLow']) || null,
          beta: parseFloat(overview.Beta) || null
        }
      };
    } catch (error) {
      logger.error(`Error fetching detailed quote for ${symbol}:`, error.message);
      throw error;
    }
  }

  async historical(symbol, options = {}) {
    try {
      const cleanSymbol = symbol.replace(/\.(NS|BSE)$/i, '');
      
      // Alpha Vantage uses TIME_SERIES_DAILY for daily data
      const response = await axios.get(ALPHA_VANTAGE_BASE_URL, {
        params: {
          function: 'TIME_SERIES_DAILY',
          symbol: cleanSymbol,
          outputsize: 'full', // Get full historical data
          apikey: this.apiKey
        }
      });

      const timeSeries = response.data['Time Series (Daily)'];
      
      if (!timeSeries) {
        throw new Error(`No historical data found for symbol: ${cleanSymbol}`);
      }

      // Convert to array format
      const historicalData = Object.entries(timeSeries).map(([date, values]) => ({
        date: new Date(date),
        open: parseFloat(values['1. open']),
        high: parseFloat(values['2. high']),
        low: parseFloat(values['3. low']),
        close: parseFloat(values['4. close']),
        volume: parseInt(values['5. volume'])
      }));

      // Sort by date ascending
      historicalData.sort((a, b) => a.date - b.date);

      // Filter by date range if provided
      if (options.period1 && options.period2) {
        return historicalData.filter(item => 
          item.date >= options.period1 && item.date <= options.period2
        );
      }

      return historicalData;
    } catch (error) {
      logger.error(`Error fetching historical data for ${symbol}:`, error.message);
      throw error;
    }
  }

  async getIntradayData(symbol, interval = '5min') {
    try {
      const cleanSymbol = symbol.replace(/\.(NS|BSE)$/i, '');
      
      const response = await axios.get(ALPHA_VANTAGE_BASE_URL, {
        params: {
          function: 'TIME_SERIES_INTRADAY',
          symbol: cleanSymbol,
          interval: interval,
          apikey: this.apiKey
        }
      });

      const timeSeries = response.data[`Time Series (${interval})`];
      
      if (!timeSeries) {
        throw new Error(`No intraday data found for symbol: ${cleanSymbol}`);
      }

      return Object.entries(timeSeries).map(([datetime, values]) => ({
        datetime: new Date(datetime),
        open: parseFloat(values['1. open']),
        high: parseFloat(values['2. high']),
        low: parseFloat(values['3. low']),
        close: parseFloat(values['4. close']),
        volume: parseInt(values['5. volume'])
      }));
    } catch (error) {
      logger.error(`Error fetching intraday data for ${symbol}:`, error.message);
      throw error;
    }
  }
}

// Create singleton instance
const alphaVantageService = new AlphaVantageService();

module.exports = alphaVantageService;

File: backend/utils/alpha_vantage_service.py
"""Alpha Vantage API service for Python"""
import requests
import os
import logging
from typing import Dict, List, Optional, Any
from datetime import datetime, timedelta

logger = logging.getLogger(__name__)

ALPHA_VANTAGE_BASE_URL = 'https://www.alphavantage.co/query'

class AlphaVantageService:
    def __init__(self):
        self.api_key = os.getenv('ALPHA_VANTAGE_API_KEY')
        if not self.api_key:
            logger.warning('ALPHA_VANTAGE_API_KEY not set in environment variables')
    
    def _clean_symbol(self, symbol: str) -> str:
        """Remove .NS or .BSE suffix from symbol"""
        return symbol.replace('.NS', '').replace('.BSE', '')
    
    def get_quote(self, symbol: str) -> Dict[str, Any]:
        """Get real-time quote for a symbol"""
        try:
            clean_symbol = self._clean_symbol(symbol)
            
            response = requests.get(ALPHA_VANTAGE_BASE_URL, params={
                'function': 'GLOBAL_QUOTE',
                'symbol': clean_symbol,
                'apikey': self.api_key
            })
            
            response.raise_for_status()
            data = response.json()
            
            quote = data.get('Global Quote', {})
            
            if not quote:
                raise ValueError(f"No data found for symbol: {clean_symbol}")
            
            return {
                'symbol': clean_symbol,
                'regularMarketPrice': float(quote.get('05. price', 0)),
                'regularMarketPreviousClose': float(quote.get('08. previous close', 0)),
                'regularMarketVolume': int(quote.get('06. volume', 0)),
                'regularMarketChange': float(quote.get('09. change', 0)),
                'regularMarketChangePercent': float(quote.get('10. change percent', '0').replace('%', '')),
                'regularMarketOpen': float(quote.get('02. open', 0)),
                'regularMarketDayHigh': float(quote.get('03. high', 0)),
                'regularMarketDayLow': float(quote.get('04. low', 0)),
                'longName': clean_symbol,
                'shortName': clean_symbol
            }
        except Exception as e:
            logger.error(f"Error fetching quote for {symbol}: {e}")
            raise
    
    def get_company_overview(self, symbol: str) -> Dict[str, Any]:
        """Get company overview including fundamentals"""
        try:
            clean_symbol = self._clean_symbol(symbol)
            
            response = requests.get(ALPHA_VANTAGE_BASE_URL, params={
                'function': 'OVERVIEW',
                'symbol': clean_symbol,
                'apikey': self.api_key
            })
            
            response.raise_for_status()
            data = response.json()
            
            # Check if we got rate limited or no data
            if not data or 'Note' in data or not data.get('Symbol'):
                return {
                    'Symbol': clean_symbol,
                    'Name': clean_symbol,
                    'MarketCapitalization': '0',
                    'PERatio': None,
                    'DividendYield': None,
                    '52WeekHigh': None,
                    '52WeekLow': None,
                    'Beta': None,
                    'Sector': 'Unknown',
                    'Industry': 'Unknown'
                }
            
            return data
        except Exception as e:
            logger.error(f"Error fetching company overview for {symbol}: {e}")
            return {
                'Symbol': self._clean_symbol(symbol),
                'Name': self._clean_symbol(symbol),
                'MarketCapitalization': '0',
                'PERatio': None,
                'DividendYield': None,
                '52WeekHigh': None,
                '52WeekLow': None,
                'Beta': None,
                'Sector': 'Unknown',
                'Industry': 'Unknown'
            }
    
    def get_detailed_quote(self, symbol: str) -> Dict[str, Any]:
        """Get comprehensive stock data combining quote and overview"""
        try:
            quote = self.get_quote(symbol)
            overview = self.get_company_overview(symbol)
            
            # Merge the data
            return {
                'symbol': quote['symbol'],
                'longName': overview.get('Name', quote['symbol']),
                'shortName': overview.get('Name', quote['symbol']),
                'regularMarketPrice': quote['regularMarketPrice'],
                'regularMarketPreviousClose': quote['regularMarketPreviousClose'],
                'regularMarketVolume': quote['regularMarketVolume'],
                'marketCap': int(float(overview.get('MarketCapitalization', 0))),
                'trailingPE': float(overview['PERatio']) if overview.get('PERatio') else None,
                'dividendYield': float(overview['DividendYield']) if overview.get('DividendYield') else None,
                'fiftyTwoWeekHigh': float(overview['52WeekHigh']) if overview.get('52WeekHigh') else None,
                'fiftyTwoWeekLow': float(overview['52WeekLow']) if overview.get('52WeekLow') else None,
                'beta': float(overview['Beta']) if overview.get('Beta') else None,
                'sector': overview.get('Sector', 'Unknown'),
                'industry': overview.get('Industry', 'Unknown')
            }
        except Exception as e:
            logger.error(f"Error fetching detailed quote for {symbol}: {e}")
            raise
    
    def get_historical_data(self, symbol: str, outputsize: str = 'compact') -> List[Dict[str, Any]]:
        """Get historical daily data"""
        try:
            clean_symbol = self._clean_symbol(symbol)
            
            response = requests.get(ALPHA_VANTAGE_BASE_URL, params={
                'function': 'TIME_SERIES_DAILY',
                'symbol': clean_symbol,
                'outputsize': outputsize,  # 'compact' (100 days) or 'full' (20+ years)
                'apikey': self.api_key
            })
            
            response.raise_for_status()
            data = response.json()
            
            time_series = data.get('Time Series (Daily)', {})
            
            if not time_series:
                raise ValueError(f"No historical data found for symbol: {clean_symbol}")
            
            # Convert to list format
            historical_data = []
            for date_str, values in time_series.items():
                historical_data.append({
                    'date': date_str,
                    'open': float(values['1. open']),
                    'high': float(values['2. high']),
                    'low': float(values['3. low']),
                    'close': float(values['4. close']),
                    'volume': int(values['5. volume'])
                })
            
            # Sort by date ascending
            historical_data.sort(key=lambda x: x['date'])
            
            return historical_data
        except Exception as e:
            logger.error(f"Error fetching historical data for {symbol}: {e}")
            raise
    
    def get_intraday_data(self, symbol: str, interval: str = '5min') -> List[Dict[str, Any]]:
        """Get intraday data"""
        try:
            clean_symbol = self._clean_symbol(symbol)
            
            response = requests.get(ALPHA_VANTAGE_BASE_URL, params={
                'function': 'TIME_SERIES_INTRADAY',
                'symbol': clean_symbol,
                'interval': interval,
                'apikey': self.api_key
            })
            
            response.raise_for_status()
            data = response.json()
            
            time_series = data.get(f'Time Series ({interval})', {})
            
            if not time_series:
                raise ValueError(f"No intraday data found for symbol: {clean_symbol}")
            
            intraday_data = []
            for datetime_str, values in time_series.items():
                intraday_data.append({
                    'datetime': datetime_str,
                    'open': float(values['1. open']),
                    'high': float(values['2. high']),
                    'low': float(values['3. low']),
                    'close': float(values['4. close']),
                    'volume': int(values['5. volume'])
                })
            
            return intraday_data
        except Exception as e:
            logger.error(f"Error fetching intraday data for {symbol}: {e}")
            raise

# Create singleton instance
alpha_vantage_service = AlphaVantageService()
